File: src/nifs/file_system.zig
const std = @import("std");
const assert = std.debug.assert;
const e = @import("../erl_nif.zig");
const rl = @import("../raylib.zig");

const core = @import("../core.zig");

pub const exported_nifs = [_]e.ErlNifFunc{
    // File system
    .{ .name = "is_file_dropped", .arity = 0, .fptr = core.nif_wrapper(nif_is_file_dropped), .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
    .{ .name = "load_dropped_files", .arity = 0, .fptr = core.nif_wrapper(nif_load_dropped_files), .flags = e.ERL_NIF_DIRTY_JOB_CPU_BOUND },
};

///////////////////
//  File system  //
///////////////////

/// Check if a file has been dropped into window
///
/// raylib.h
/// RLAPI bool IsFileDropped(void);
fn nif_is_file_dropped(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) !e.ErlNifTerm {
    assert(argc == 0);
    _ = argv;

    // Function

    const is_file_dropped = rl.IsFileDropped();

    // Return

    return core.Boolean.make(env, is_file_dropped);
}

/// Load dropped filepaths
///
/// raylib.h
/// RLAPI FilePathList LoadDroppedFiles(void);
fn nif_load_dropped_files(env: ?*e.ErlNifEnv, argc: c_int, argv: [*c]const e.ErlNifTerm) !e.ErlNifTerm {
    assert(argc == 0);
    _ = argv;

    // Function

    const dropped_files = rl.LoadDroppedFiles();
    defer rl.UnloadDroppedFiles(dropped_files);

    // Return

    const paths_lengths = [_]usize{ @intCast(dropped_files.count), @intCast(rl.MAX_FILEPATH_LENGTH) };
    return core.Array.make_c(core.CString, [*c]u8, env, dropped_files.paths, &paths_lengths);
}

File: lib/zexray/nif/file_system.ex
defmodule Zexray.NIF.FileSystem do
  @moduledoc false

  defmacro __using__(_opts) do
    quote do
      @nifs_file_system [
        # File system
        is_file_dropped: 0,
        load_dropped_files: 0
      ]

      #################
      #  File system  #
      #################

      @doc """
      Check if a file has been dropped into window

      ```c
      // raylib.h
      RLAPI bool IsFileDropped(void);
      ```
      """
      @doc group: :file_system
      @spec is_file_dropped() :: boolean
      def is_file_dropped(), do: :erlang.nif_error(:undef)

      @doc """
      Load dropped filepaths

      ```c
      // raylib.h
      RLAPI FilePathList LoadDroppedFiles(void);
      ```
      """
      @doc group: :file_system
      @spec load_dropped_files() :: [binary]
      def load_dropped_files(), do: :erlang.nif_error(:undef)
    end
  end
end
